src/mds/Migrator.cc

```diff
@@ -242,6 +242,20 @@ void Migrator::find_stale_export_freeze()
   }
 }
 
+void Migrator::quiesce_overdrive_export(CDir *dir) {
```

Comment:
auth pinning issue:

```
2024-05-20T18:18:38.226+0000 7f2a8b269700  1 -- [v2:172.21.10.4:6834/3955013195,v1:172.21.10.4:6835/3955013195] <== mds.0 v2:172.21.10.4:6832/3457661485 400 ==== export_discover_ack(0x20000000001 success) ==== 13+0+0 (crc 0 0 0) 0x5564cccf21a0 con 0x5564ccccd200
2024-05-20T18:18:38.226+0000 7f2a84a5c700 10 MDSIOContextBase::complete: 26C_Locker_FileUpdate_finish
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.mig handle_export_discover_ack from mds.0 on [dir 0x20000000001 /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ [2,head] auth{0=1} pv=1510 v=1505 cv=1458/1458 dir_auth=1 ap=5+11 state=1610875909|complete|freezingtree|exporting f(v4 m2024-05-20T18:18:36.930257+0000 207=207+0)/f(v4 m2024-05-20T18:18:35.368297+0000 206=206+0)->f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v0 rc2024-05-20T18:18:36.965047+0000 b22026675 204=204+0)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 199=199+0)->n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 204=204+0) hs=207+1,ss=0+0 dirty=16 | ptrwaiter=1 request=1 child=1 frozen=0 subtree=1 exportbound=0 replicated=1 dirty=1 waiter=1 authpin=1 0x5564cce9cd00]
2024-05-20T18:18:38.226+0000 7f2a8b269700 15 mds.1.526 hit export target is [C 6.49e+01]
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.cache request_finish request(mds.1:34 nref=3)
2024-05-20T18:18:38.226+0000 7f2a8b269700 15 mds.1.cache request_cleanup request(mds.1:34 nref=3)
2024-05-20T18:18:38.226+0000 7f2a8425b700  5 mds.1.log _submit_thread 117404609~12202 : ESubtreeMap 9 subtrees , 0 ambiguous [metablob 0x101, 15 dirs]
2024-05-20T18:18:38.226+0000 7f2a8b269700  1 -- [v2:172.21.10.4:6834/3955013195,v1:172.21.10.4:6835/3955013195] send_to--> mds [v2:172.21.10.4:6832/3457661485,v1:172.21.10.4:6833/3457661485] -- peer_request(mds.1:34.0 finish) -- ?+0 0x5564ceaa5180
2024-05-20T18:18:38.226+0000 7f2a8b269700  1 -- [v2:172.21.10.4:6834/3955013195,v1:172.21.10.4:6835/3955013195] --> [v2:172.21.10.4:6832/3457661485,v1:172.21.10.4:6833/3457661485] -- peer_request(mds.1:34.0 finish) -- 0x5564ceaa5180 con 0x5564ccccd200
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.locker rdlock_finish on (isnap sync r=234) on [inode 0x1 [...2,head] / rep@0.1 v8 snaprealm=0x5564cce34240 f(v0 m2024-05-20T18:18:08.789396+0000 1=0+1) n(v2 rc2024-05-20T18:18:19.293889+0000 b133 7=2+5)/n(v0 rc2024-05-20T18:17:57.639634+0000 1=0+1) (isnap sync r=234) (inest mix) | lock=1 dirfrag=1 dirwaiter=0 discoverbase=0 0x5564ccddec00]
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.locker rdlock_finish on (isnap sync r=234) on [inode 0x10000000000 [...2,head] /volumes/ rep@0.1 v13 f(v0 m2024-05-20T18:18:08.820181+0000 2=1+1) n(v1 rc2024-05-20T18:18:08.820181+0000 b133 6=2+4)/n(v0 rc2024-05-20T18:18:08.820181+0000 b133 6=2+4) (isnap sync r=234) (inest mix) (ifile lock) caps={20116=pAsLsXs/-@2,30071=pAsLsXs/-@2} | request=0 lock=1 dirfrag=1 caps=1 waiter=0 0x5564ccddf180]
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.locker rdlock_finish on (isnap sync r=234) on [inode 0x10000000001 [...2,head] /volumes/_nogroup/ rep@0.1 v39 f(v0 m2024-05-20T18:18:08.795245+0000 1=0+1) n(v3 rc2024-05-20T18:18:36.840136+0000 b340327133 3666=3603+63)/n(v0 rc2024-05-20T18:18:08.818620+0000 b133 4=1+3) (isnap sync r=234) (inest mix) (ifile lock) caps={20116=pAsLsXs/-@2,30071=pAsLsXs/-@2} | request=0 lock=1 dirfrag=1 caps=1 waiter=0 0x5564ccddf700]
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.locker rdlock_finish on (isnap sync r=234) on [inode 0x10000000002 [...2,head] /volumes/_nogroup/subvol_quiesce/ rep@0.1 v40 snaprealm=0x5564cce34d80 f(v0 m2024-05-20T18:18:08.818620+0000 2=1+1) n(v2 rc2024-05-20T18:18:31.897630+0000 b150762321 1624=1572+52)/n(v0 rc2024-05-20T18:18:08.818620+0000 b133 3=1+2) (isnap sync r=234) (inest mix) (ifile lock) caps={20116=pAsLsXs/-@2,30071=pAsLsXs/-@2} | request=0 lock=1 dirfrag=1 caps=1 waiter=0 0x5564cce1f080]
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.locker rdlock_finish on (isnap sync r=234) on [inode 0x10000000003 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/ rep@0.1 v183 f(v0 m2024-05-20T18:18:19.595920+0000 20=0+20) n(v2 rc2024-05-20T18:18:37.417075+0000 b352501851 3821=3760+61)/n(v0 rc2024-05-20T18:18:08.807923+0000 1=0+1) (isnap sync r=234) (inest mix) (ipolicy sync r=122) (iquiesce lock x=1 by request(mds.1:2 nref=4)) caps={20116=pAsLsXs/-@4,30071=pAsLsXs/-@6} | request=1 lock=3 dirfrag=1 caps=1 dirwaiter=0 waiter=0 discoverbase=0 0x5564cce1f600]
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.locker rdlock_finish on (isnap sync) on [inode 0x100000005e2 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/ rep@0.1 v379 RANDEPHEMERALPIN f(v1 m2024-05-20T18:18:24.538791+0000 2=1+1) n(v2 rc2024-05-20T18:18:27.326550+0000 b13305413 203=200+3)/n(v0 rc2024-05-20T18:18:19.416979+0000 2=1+1) (inest mix) (ipolicy sync r=1) (iquiesce lock x=1 by request(mds.1:9 nref=4)) caps={30071=pAsLsXs/-@1} | dirtyscattered=0 request=1 lock=2 importing=0 dirfrag=1 caps=1 waiter=0 randepin 0x5564cce20680]
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.locker local_wrlock_finish  on (iquiesce lock w=1) on [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ifile mix w=1 scatter_wanted) (iquiesce lock w=1) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=5 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700 10 MDSContext::complete: 18C_MDS_RetryRequest
2024-05-20T18:18:38.226+0000 7f2a8b269700 20 mds.1.cache dispatch_quiesce_inode request(mds.1:31 nref=2) quiescing [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ifile mix w=1 scatter_wanted) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=4 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700 10 mds.1.locker acquire_locks request(mds.1:31 nref=2)
2024-05-20T18:18:38.226+0000 7f2a8b269700 20 mds.1.locker  lov = [LockOp(l=(iquiesce lock),f=0x4),LockOp(l=(ipolicy sync),f=0x1)]
2024-05-20T18:18:38.226+0000 7f2a8b269700 20 mds.1.locker  auth_pin_nonblocking=0
2024-05-20T18:18:38.226+0000 7f2a8b269700 20 mds.1.locker  must xlock (iquiesce lock) [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ifile mix w=1 scatter_wanted) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=4 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700 20 mds.1.locker  must rdlock (ipolicy sync) [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ifile mix w=1 scatter_wanted) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=4 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700 10 mds.1.locker  must authpin [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ifile mix w=1 scatter_wanted) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=4 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700 10 mds.1.locker  already auth_pinned [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ifile mix w=1 scatter_wanted) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=4 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.locker local_xlock_start  on (iquiesce lock) on [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ifile mix w=1 scatter_wanted) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=4 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700 10 mds.1.locker  got xlock on (iquiesce lock x=1 by request(mds.1:31 nref=5)) [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ifile mix w=1 scatter_wanted) (iquiesce lock x=1 by request(mds.1:31 nref=5)) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=5 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700  7 mds.1.locker rdlock_start  on (ipolicy sync) on [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ifile mix w=1 scatter_wanted) (iquiesce lock x=1 by request(mds.1:31 nref=5)) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=5 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700 10 mds.1.locker  got rdlock on (ipolicy sync r=1) [inode 0x20000000001 [...2,head] /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ rep@0.1 v1151 RANDEPHEMERALPIN f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 205=204+1)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 200=199+1) (idft sync r=1) (isnap sync r=1) (inest mix w=1 scatter_wanted) (ipolicy sync r=1) (ifile mix w=1 scatter_wanted) (iquiesce lock x=1 by request(mds.1:31 nref=5)) caps={30071=pAsLsXs/-@6} | ptrwaiter=0 dirtyscattered=0 request=3 lock=6 dirfrag=1 caps=1 exportingcaps=0 dirtyparent=0 replicated=0 dirty=0 waiter=1 authpin=0 tempexporting=0 randepin 0x5564ccf15180]
2024-05-20T18:18:38.226+0000 7f2a8b269700 10 mds.1.mig quiesce_overdrive_export quiesce_overdrive_exportwill try to cancel in state: (3) discovering
2024-05-20T18:18:38.226+0000 7f2a8b269700 10 mds.1.mig export_try_cancel [dir 0x20000000001 /volumes/_nogroup/subvol_quiesce/b4bdffb8-5abd-4904-aa79-f16fcaecebcc/tmp.lOHRSm6Cyc/usr/bin/ [2,head] auth{0=1} pv=1510 v=1505 cv=1458/1458 dir_auth=1 ap=5+11 state=1610875909|complete|freezingtree|exporting f(v4 m2024-05-20T18:18:36.930257+0000 207=207+0)/f(v4 m2024-05-20T18:18:35.368297+0000 206=206+0)->f(v5 m2024-05-20T18:18:36.930257+0000 207=207+0) n(v0 rc2024-05-20T18:18:36.965047+0000 b22026675 204=204+0)/n(v0 rc2024-05-20T18:18:27.326550+0000 b13305413 199=199+0)->n(v1 rc2024-05-20T18:18:36.965047+0000 b22026675 204=204+0) hs=207+1,ss=0+0 dirty=16 | ptrwaiter=1 request=1 child=1 frozen=0 subtree=1 exportbound=0 replicated=1 dirty=1 waiter=1 authpin=1 0x5564cce9cd00]
2024-05-20T18:18:38.226+0000 7f2a8b269700 10 mds.1.mig export_try_cancel export state=discovering : canceling freeze and removing auth_pin
```

issue is:

https://github.com/ceph/ceph/blob/acdd6bc00bfd8e7bf10533befba15ce193d11b90/src/mds/Migrator.cc#L1234

but that resumes the quiesce op which triggers an export cancel. However! The export state should be `EXPORT_FREEZING` but `handle_export_discover_ack` has not reached that point yet. This causes us to double `dir->auth_unpin`. Once here:

https://github.com/ceph/ceph/blob/acdd6bc00bfd8e7bf10533befba15ce193d11b90/src/mds/Migrator.cc#L264

and then

https://github.com/ceph/ceph/blob/acdd6bc00bfd8e7bf10533befba15ce193d11b90/src/mds/Migrator.cc#L1238

The fix I think is to set the export state before finishing the request.

Reply:
ah dang! I think we just need to make that overdrive async

Reply:
or yes, changing the order of the state set

Reply:
These issues are now tackled in https://github.com/ceph/ceph/pull/57579